trieste/acquisition/rule.py

```diff
@@ -37,86 +38,72 @@
     SingleModelAcquisitionBuilder,
     SingleModelGreedyAcquisitionBuilder,
 )
-from .optimizer import (
-    AcquisitionOptimizer,
-    automatic_optimizer_selector,
-    batchify,
-    generate_continuous_optimizer,
-)
+from .optimizer import AcquisitionOptimizer, automatic_optimizer_selector, batchify
 from .sampler import ExactThompsonSampler, RandomFourierFeatureThompsonSampler, ThompsonSampler
 
-S = TypeVar("S")
-""" Unbound type variable. """
+T_co = TypeVar("T_co", covariant=True)
+""" Unbound covariant type variable. """
```

Comment:
What does this do?

Reply:
if `Foo` is parametrized by a **covariant** type variable (as `F[T_co]`), then `Foo[Child]` is a subclass of `Foo[Parent]`
if `Foo` is parametrized by a **contravariant** type variable (as `F[T_contra]`), then `Foo[Parent]` is a subclass of `Foo[Child]`
if `Foo` is parametrized by an invariant type variable (as `F[T]`), then neither of `Foo[Child]` or `Foo[Parent]` is a subclass of the other

In python, it's standard to suffix co- and contravariant type variables by `_co` and `_contra`

Reply:
this means that an `AcquisitionRule[int]` is a subclass of `AcquisitionRule[float]` because `int` is (at type checking at least), a subclass of `float`

Reply:
specifically, that I can use an `AcquisitionRule[int]` when an `AcquisitionRule[float]` is required. Or perhaps more relevantly, I can use an `AcquisitionRule[tf.Tensor]` where an `AcquisitionRule[TensorType]` is needed because a `tf.Tensor` is a `tf.Tensor | np.ndarray`.

Reply:
note it's not needed for this PR, but it seems like good form

Reply:
Thanks for explaining!